lib/experimental/interactivity-api/directive-processing.php

```diff
@@ -28,37 +28,255 @@ function gutenberg_interactivity_mark_root_blocks( $parsed_block, $source_block,
 add_filter( 'render_block_data', 'gutenberg_interactivity_mark_root_blocks', 10, 3 );
 
 /**
- * Process directives in each root block.
+ * Processes the directives in the root blocks.
  *
  * @param string $block_content The block content.
  * @param array  $block         The full block.
  *
  * @return string Filtered block content.
  */
 function gutenberg_process_directives_in_root_blocks( $block_content, $block ) {
+	static $directives = array(
+		'data-wp-bind'    => 'gutenberg_interactivity_process_wp_bind',
+		'data-wp-context' => 'gutenberg_interactivity_process_wp_context',
+		'data-wp-class'   => 'gutenberg_interactivity_process_wp_class',
+		'data-wp-style'   => 'gutenberg_interactivity_process_wp_style',
+		'data-wp-text'    => 'gutenberg_interactivity_process_wp_text',
+	);
+
 	if ( WP_Directive_Processor::is_marked_as_root_block( $block ) ) {
 		WP_Directive_Processor::unmark_root_block();
-		$directives = array(
-			'data-wp-bind'    => 'gutenberg_interactivity_process_wp_bind',
-			'data-wp-context' => 'gutenberg_interactivity_process_wp_context',
-			'data-wp-class'   => 'gutenberg_interactivity_process_wp_class',
-			'data-wp-style'   => 'gutenberg_interactivity_process_wp_style',
-			'data-wp-text'    => 'gutenberg_interactivity_process_wp_text',
-		);
 
-		$tags = new WP_Directive_Processor( $block_content );
-		$tags = $tags->process_rendered_html( $tags, 'data-wp-', $directives );
-		return $tags->get_updated_html();
+		// Parse our own block delimiters for interactive and non-interactive blocks.
+		$parsed_blocks     = parse_blocks( $block_content );
+		$context           = new WP_Directive_Context();
+		$processed_content = '';
 
+		foreach ( $parsed_blocks as $parsed_block ) {
+			if ( 'core/interactivity-wrapper' === $parsed_block['blockName'] ) {
+				$processed_content .= gutenberg_process_interactive_block( $parsed_block, $context, $directives );
+			} elseif ( 'core/non-interactivity-wrapper' === $parsed_block['blockName'] ) {
+				$processed_content .= gutenberg_process_non_interactive_block( $parsed_block, $context, $directives );
```

Comment:
The code is never passing through this conditional when it should.

Reply:
I think I have found the issue! Probably, we shouldn't remove anything from `$children_of_interactive_block`.

Reply:
This is a bit tricky to explain. The thing is that the code that marks a block as a child of an interactive block runs in `render_block_data`, and that hook is dispatched during the block's parent render. ☝️ 

When the `render.php` file of an interactive block executes, the following hooks have already been dispatched for each **inner block**:
- `render_block_data` (adds the block to `$children_of_interactive_block`)
- `render_block` (remove the block from `$children_of_interactive_block`)

That means inner blocks have already been processed, have their interactive/non-interactive wrappers, and, more importantly, have been removed from `$children_of_interactive_block`.

If we now overwrite `$content` and run `$inner_block->render()` for each inner block,  **only `render_block` is dispatched** this time, so the code fails to check whether the block is a child of an interactive block.

I guess we could simply leave all children's IDs in `$children_of_interactive_block` to avoid this problem, right? 🤔

Reply:
I've seen @c4rl0sbr4v0 has removed the call to `unmark_children_of_interactive_block ` in https://github.com/WordPress/gutenberg/pull/56302/commits/bef900b27ef0aae402f5138ea688a167edce2fe6. 👍 

@luisherranz, as you modified that method in https://github.com/WordPress/gutenberg/pull/56302/commits/276b64cef7aeb60be77f92a5277597174743a376, what do you think of what I've posted above, in https://github.com/WordPress/gutenberg/pull/56302#discussion_r1434595284? Does it make sense?

Reply:
If we want the current approach to work with `$inner_block->render()` calls, I guess we need to find a way to unmark the processed children in the parent, i.e., during the `render_block` of the interactive block. Any ideas? 🤔

Reply:
I think something like this should be enough @c4rl0sbr4v0 @luisherranz 

```diff
diff --git a/lib/experimental/interactivity-api/directive-processing.php b/lib/experimental/interactivity-api/directive-processing.php
index 7944bfb84d..73e9826700 100644
--- a/lib/experimental/interactivity-api/directive-processing.php
+++ b/lib/experimental/interactivity-api/directive-processing.php
@@ -99,6 +99,9 @@ function gutenberg_mark_block_interactivity( $block_content, $block, $block_inst
 		isset( $block_instance->block_type->supports['interactivity'] ) &&
 		$block_instance->block_type->supports['interactivity']
 	) {
+		foreach ( $block['innerBlocks'] as $inner_block ) {
+			WP_Directive_Processor::unmark_children_of_interactive_block( $inner_block );
+		}
 		// Wraps the interactive block with a comment delimiter to be able to
 		// process it later.
 		return get_comment_delimited_block_content(
```

Reply:
Unmarking children during the interactive block rendering could add extraneous wrappers for duplicated blocks, like in the image below, but I think it's not critical, and it supports inner block re-rendering.

![image](https://github.com/WordPress/gutenberg/assets/6917969/c6d3fa82-a6fe-47b1-acb1-979ebc37d747)

Reply:
On the other hand, if we unmark children during their rendering, the re-render of inner blocks inside an interactive block is not supported.

![image](https://github.com/WordPress/gutenberg/assets/6917969/ae94c4ae-652f-49ca-864a-7d510f112b7b)

Reply:
Even with the extra wrappers, I believe it's safe to unmark children of interactive blocks during the interactive block's rendering, keeping in mind that those extra wrappers would always be `<!-- core/non-interactive-wrapper -->`, and there's no processing at all for these kind of blocks.

cc: @luisherranz

Reply:
There's not much gain in unmarking those blocks, is there?

I guess Carlos added it to keep things clean, but it's not strictly required. I changed the logic a bit to remove only one md5 and not all the md5's because, with the md5 technique, multiple blocks can have the same id (md5), but it's not very important either.

Once we get rid of the md5s, we can mark the block in the `$parsed_block` with an extra property (`$parsed_block['is_children...'] = true;`) and never remove it, so this problem will disappear.

For now, I'd just remove the unmark logic.

By the way:

- It's not the end of the world if there are some edge cases where the non-interactive blocks are processed, as long as those don't happen very often. All this is an optimization, and optimizations don't need to be perfect. The important bit is that _all_ the interactive blocks are processed.
- There is a `skip_inner_blocks` property that blocks should use when they are calling the `->render()` method during their render [to avoid rendering the inner blocks twice](https://github.com/WordPress/wordpress-develop/blob/c973689ff491317b63bcafb8fefcc70473a9587d/src/wp-includes/class-wp-block.php#L220).

Reply:
I made a unit test to test this out and it still fails with `skip_inner_blocks` with either the original `WP_Directive_Processor::unmark_children_of_interactive_block( $block );` or with David's fix, so that doesn't help.

Even though David's fix is the best way to unmark the blocks, I've removed that code to keep things as simple as possible, taking into account that unmarking doesn't have any benefit, and we won't do it once we get rid of the md5s.